File: 2973_find_number_of_coins_to_place_in_tree_nodes/solution.go
// LeetCode 2973 - Find Number of Coins to Place in Tree Nodes
// https://leetcode.com/problems/find-number-of-coins-to-place-in-tree-nodes/

import "sort"

func placedCoins(edges [][]int, cost []int) []int64 {
	n := len(cost)
	g := make([][]int, n)
	for _, e := range edges {
		u, v := e[0], e[1]
		g[u] = append(g[u], v)
		g[v] = append(g[v], u)
	}
	ans := make([]int64, n)
	var dfs func(int, int) []int
	dfs = func(u, p int) []int {
		vals := []int{cost[u]}
		for _, v := range g[u] {
			if v == p {
				continue
			}
			vals = append(vals, dfs(v, u)...)
		}
		sort.Ints(vals)
		if len(vals) < 3 {
			ans[u] = 1
		} else {
			m := len(vals)
			cand1 := int64(vals[m-1]) * int64(vals[m-2]) * int64(vals[m-3])
			cand2 := int64(vals[0]) * int64(vals[1]) * int64(vals[m-1])
			best := cand1
			if cand2 > best {
				best = cand2
			}
			if best < 0 {
				best = 0
			}
			ans[u] = best
		}
		if len(vals) <= 5 {
			return vals
		}
		return append(append([]int{}, vals[:2]...), vals[len(vals)-3:]...)
	}
	dfs(0, -1)
	return ans
}

File: 2973_find_number_of_coins_to_place_in_tree_nodes/solution.rb
# LeetCode 2973 - Find Number of Coins to Place in Tree Nodes
# https://leetcode.com/problems/find-number-of-coins-to-place-in-tree-nodes/

# @param {Integer[][]} edges
# @param {Integer[]} cost
# @return {Integer[]}
def placed_coins(edges, cost)
  n = cost.length
  g = Array.new(n) { [] }
  edges.each do |e|
    g[e[0]] << e[1]
    g[e[1]] << e[0]
  end
  ans = Array.new(n, 0)
  dfs = lambda do |u, p|
    vals = [cost[u]]
    g[u].each do |v|
      next if v == p

      vals += dfs.call(v, u)
    end
    vals.sort!
    if vals.length < 3
      ans[u] = 1
    else
      m = vals.length
      cand1 = vals[m - 1] * vals[m - 2] * vals[m - 3]
      cand2 = vals[0] * vals[1] * vals[m - 1]
      best = [cand1, cand2].max
      best = 0 if best < 0
      ans[u] = best
    end
    return vals if vals.length <= 5

    [vals[0], vals[1], vals[-3], vals[-2], vals[-1]]
  end
  dfs.call(0, -1)
  ans
end
